numpy/core/src/multiarray/datetime_strings.c

```diff
@@ -1434,24 +1436,27 @@ array_datetime_as_string(PyObject *NPY_UNUSED(self), PyObject *args,
 
     /* Get the input time zone */
     if (timezone_obj != NULL) {
-        /* Convert to ASCII if it's unicode */
-        if (PyUnicode_Check(timezone_obj)) {
-            /* accept unicode input */
-            PyObject *obj_str;
-            obj_str = PyUnicode_AsASCIIString(timezone_obj);
+        PyObject *strobj;
+        if (PyBytes_Check(timezone_obj)) {
+            /* accept bytes input */
+            PyObject *obj_str = PyUnicode_FromEncodedObject(timezone_obj, NULL, NULL);
             if (obj_str == NULL) {
                 goto fail;
             }
-            Py_DECREF(timezone_obj);
-            timezone_obj = obj_str;
+            strobj = obj_str;
         }
+        else {
+            Py_INCREF(timezone_obj);
+            strobj = timezone_obj;
+        }
+
+        Py_SETREF(timezone_obj, strobj);
```

Comment:
This `SETREF` feels a bit weird, since it is just `strobj`. But I guess you wanted to keep the unicode-convert code snippet is always identical?

Reply:
> But I guess you wanted to keep the unicode-convert code snippet is always identical?

Yeah, I think that was my thought. I'd considered extracting it to a helper function a some point, likely bundling a deprecation warning because there's no reason to accept byte-strings here anyway.